### ISSUE TITLE
Feature Request: A way to pass data between authenticator (and request handler) contexts

### ISSUE THREAD

**Issue by colstrom** [hirura/hrr_rb_ssh#7]
Splitting this out from #5 as requested.

The reference handlers for `pty_req` and `shell` use `context.variables` to pass data from one handler to the next (in that case, `pts` and `ptm` created in the `pty_req` handler and used in the `shell` handler).

Is there a way to pass data between authenticators (and also from authenticators to request_handlers) in a similar manner? I'm thinking of two scenarios here...

First, passing between authenticators would allow for things like requiring both public key and password authentication, or other multi-step auth flows.

Secondly, passing between authenticators and request_handlers would allow for things like creating an encryption context from the user's password, and using that for encrypting/decrypting things like stored secrets.

**Comment by hirura** [comment]
Thank you opening the issue.

Passing data in also authenticators is very interesting idea. The current implementation only passes username to subsequent channel-request handlers. As @colstrom mentions, the two scenarios would be useful. In addition, keyboard-interactive authentication would be more useful to handling additional data.

I will try to implement such feature.

**Comment by hirura** [comment]
I have implemented authenticator's `Context#variables` that is shared among authenticators and channel request handlers. Please refer to #8 and try [v0.3.0.pre2](https://rubygems.org/gems/hrr_rb_ssh/versions/0.3.0.pre2).

**Comment by colstrom** [comment]
Fantastic. Using `0.3.0.pre2`, I've been able to perform a two-step authentication by setting `context.variables[:authenticated_with_public_key] = true` and returning `false` from a public_key authenticator, and using a `return false unless context.variables[:authenticated_with_public_key]` at the top of my password authenticator.

I've also confirmed that these are passed to the request_handler. Looks good so far!
